lib/Runtime/Library/JavascriptPromise.cpp

```diff
@@ -559,6 +559,147 @@ namespace Js
+                thenFinally = library->CreatePromiseThenFinallyFunction(EntryThenFinallyFunction, RecyclableObject::FromVar(args[1]), constructor, false);
+                catchFinally = library->CreatePromiseThenFinallyFunction(EntryThenFinallyFunction, RecyclableObject::FromVar(args[1]), constructor, true);
+            }
+            else
+            {
+                thenFinally = args[1];
+                catchFinally = args[1];
+            }
+        }
+        else
+        {
+            thenFinally = library->GetUndefined();
+            catchFinally = library->GetUndefined();
+        }
+
+        //7. Return ? Invoke(promise, "then", « thenFinally, catchFinally »).
+        Var funcVar = JavascriptOperators::GetProperty(promise, Js::PropertyIds::then, scriptContext);
+        if (!JavascriptConversion::IsCallable(funcVar))
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_FunctionArgument_NeedFunction, _u("Promise.prototype.finally"));
+        }
+        RecyclableObject* func = RecyclableObject::UnsafeFromVar(funcVar);
+
+        return CALL_FUNCTION(scriptContext->GetThreadContext(),
+            func, Js::CallInfo(CallFlags_Value, 3),
+            promise,
+            thenFinally,
+            catchFinally);
+    }
+
+    //ThenFinallyFunction as described in draft of ES2018  section 25.4.5.3.1
+    //AND CatchFinallyFunction as described in draft of ES2018  section 25.4.5.3.2
+    Var JavascriptPromise::EntryThenFinallyFunction(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+        ScriptContext* scriptContext = function->GetScriptContext();
+        
+        JavascriptLibrary* library = scriptContext->GetLibrary();
+
+        JavascriptPromiseThenFinallyFunction* This = JavascriptPromiseThenFinallyFunction::FromVar(function);
+
+        //1. Let onFinally be F.[[OnFinally]]
+        //2. Assert: IsCallabale(onFinally)=true
+        Assert(JavascriptConversion::IsCallable(This->GetOnFinally()));
+
+        //3. Let result be ? Call(onFinally, undefined)
+        Var result = CALL_FUNCTION(scriptContext->GetThreadContext(), This->GetOnFinally(), CallInfo(CallFlags_Value, 1), library->GetUndefined());
+
+        //4. Let C be F.[[Constructor]]
+        //5.Assert IsConstructor(C)
+        Assert(JavascriptOperators::IsConstructor(This->GetConstructor()));
+
+        //6. Let promise be ? PromiseResolve(c, result)
+        Var promiseVar = CreateResolvedPromise(result, scriptContext, This->GetConstructor());
+
+        //7. Let valueThunk be equivalent to a function that returns value
+        //OR 7. Let thrower be equivalent to a function that throws reason
+        JavascriptPromiseThunkFinallyFunction* Thunk = library->CreatePromiseThunkFinallyFunction(EntryThunkFinallyFunction, args[1], This->GetShouldThrow());
```

Comment:
Same naming nitpick as above. Can we not use `Thunk` here but something like `thunkFinallyFunction`?

Reply:
changed